### ISSUE TITLE
custom webpack plugin

### ISSUE THREAD

**Issue by veeramarni** [sysgears/spin.js#12]
To be honest `spin.js` works very well and most of it comes out of the box. It frees up a lot of build configuration time, but what if we want to overwrite the out of the box webpack configuration or add new webpack plugin. For example like I was referring in #10 which needs `css` plugin. 

I do see you have `plugins` option, was there any example how to use it. Will this be the way to add custom `webpack plugins` based on our needs? I don't feel like I need any additional plugins other than that comes from out of the box from `spin.js` right now but want to make sure this is supported as well.

**Comment by vlasenko** [comment]
@veeramarni There are multiple ways to do it, but I'm going to support `css` out of the box.

To customize webpack config you can use `webpackAppConfig.js`, just add proper rule to load `css` as you normally do with webpack. This approach work, if your rule is something that doesn't interfere with config generated by `spin.js`

You can also write `spin.js` if you need clever webpack config part generation, the way `spin.js` does it now. There is no docs for writing a plugin yet, but you can check source code of `spin.js` now. All webpack config generation stuff is done via plugins.

The general direction of `spin.js` is to provide as much out of the box as possible and provide mutliple ways to configure stuff, when out of the box can't work for some reason of another. Generally we want `spin.js` to not stay on developers way. We are going to add `spin init` command soon that will help you bootstrap small projects instantly by generating `package.json` with proper dependencies, after that we are going to work on supporting micro-services in the best way that can be possibly achieved.

**Comment by veeramarni** [comment]
Thanks for quick explanation. How do we overwrite few configuration, for example

If I do a production build, I get the following error


```
[at-loader] Ok, 0.446 sec.
webpack-for-server Time: 4672ms

ERROR in index.js from UglifyJs
Unexpected token: name (CounterRepository) [index.js:337,4]

[at-loader] Checking started in a separate process...

[at-loader] Ok, 0.674 sec.
webpack-for-web Time: 9285ms

ERROR in index.889b365ddea4346c3b4e.js from UglifyJs
Unexpected token operator «=», expected punc «,» [index.889b365ddea4346c3b4e.js:104,24]

ERROR in vendor.889b365ddea4346c3b4e.js from UglifyJs
Unexpected token: name (ClientLogger) [vendor.889b365ddea4346c3b4e.js:51881,6]
```

It was known issue with `uglify` that comes with web pack as it has some compilation issue with `typescript` and `ES2015` as referred [here](https://github.com/webpack-contrib/uglifyjs-webpack-plugin/issues/105). I need to use latest `uglify` to fix this issue like shown below.  

load 
```
    "uglifyjs-webpack-plugin": "^1.0.0-beta.2",
```

And then in webpack config.js
```
const UglifyJSPlugin = require('uglifyjs-webpack-plugin')
...
if (__DEV__) {
    basePlugins.push(new webpack.NamedModulesPlugin());
} else {
    basePlugins.push(new UglifyJSPlugin());
    basePlugins.push(new webpack.LoaderOptionsPlugin({ minimize: true }));
    basePlugins.push(new webpack.optimize.ModuleConcatenationPlugin());
}
```

How do we do this fix?

**Comment by vlasenko** [comment]
@veeramarni I propose to have the code in `spin.js` that detects if `uglifyjs-webpack-plugin` is in package.json `devDependencies` and if it is, then `spin.js` will use it instead of stock `UglifyJSPlugin`

**Comment by vlasenko** [comment]
@veeramarni I think if you just install 
``` json
    "uglifyjs-webpack-plugin": "^1.0.0-beta.2",
```
it should be used automatically without any changes to `spin.js`

**Comment by veeramarni** [comment]
In my case, it is already installed. But we need to add that plugin to `webpack-confg` to use it.

**Comment by vlasenko** [comment]
@veeramarni But this plugin is added to config, during build.

**Comment by veeramarni** [comment]
@vlasenko can you please elaborate? Do you mean i need to add something to `webpackAppConfig` file.

**Comment by vlasenko** [comment]
@veeramarni You need to just add the version of `uglifyjs-webpack-plugin` you want into `devDependencies` of your project:
```
yarn add -D uglifyjs-webpack-plugin@^1.0.0-beta.2
```

And this version will be used by `webpack` automatically.

**Comment by vlasenko** [comment]
https://github.com/sysgears/spin.js/blob/master/src/plugins/WebpackPlugin.ts#L33
https://github.com/webpack/webpack/blob/master/lib/optimize/UglifyJsPlugin.js#L7

**Comment by veeramarni** [comment]
I do have it in the `devDependencies`. 

But `Webpack` brings the version it is compatible with.
https://github.com/webpack/webpack/blob/master/package.json#L25

In my case, I have two `uglifyjs-webpack-plugin` and it uses the older version by default. 

![image](https://user-images.githubusercontent.com/3382355/31645657-40b983fc-b2cb-11e7-97a5-23ca34e5f505.png)
![image](https://user-images.githubusercontent.com/3382355/31645668-4c0ef58e-b2cb-11e7-8d9e-b43c71fe6cf3.png)

**Comment by vlasenko** [comment]
@veeramarni Strange that it doesn't behave like that for me. Okay, I have patched `spin.js` to use `uglifyjs-webpack-plugin` directly:
https://github.com/sysgears/spin.js/commit/95a07b16fc71e15f8fe90095f5f0bbd7269c05dc
Please try `spinjs@0.4.12`

**Comment by veeramarni** [comment]
thanks for quick fix, but getting the following error

```
> cross-env NODE_ENV=production spin build

spin error { [Error: Validation Error

UglifyJs Plugin Invalid Options

options['minimize'] should NOT have additional properties
]
  err:
   [ { keyword: 'additionalProperties',
       dataPath: '[\'minimize\']',
       schemaPath: '#/additionalProperties',
       params: { additionalProperty: 'minimize' },
       message: 'should NOT have additional properties' } ] }
```

**Comment by vlasenko** [comment]
@veeramarni Should be fixed in `spinjs@0.4.13`
